File: src/main/plans.ts
import { ipcMain } from "electron";
import { readFile, writeFile } from "fs/promises";
import { join } from "path";
import type { Plan } from "../shared/types";
import { getMemoryDir, PLANS_FILE } from "./memory";
import { generateId, fileExists } from "./utils";

function getPlansPath(): string {
  return join(getMemoryDir(), PLANS_FILE);
}

function getPlansPathIn(memoryDir: string): string {
  return join(memoryDir, PLANS_FILE);
}

export async function loadPlansIn(memoryDir: string): Promise<Record<string, Plan>> {
  const path = getPlansPathIn(memoryDir);
  if (!(await fileExists(path))) return {};
  const data = JSON.parse(await readFile(path, "utf-8"));
  return typeof data === "object" && data !== null ? data : {};
}

async function loadPlans(): Promise<Record<string, Plan>> {
  return loadPlansIn(getMemoryDir());
}

export async function savePlansIn(memoryDir: string, plans: Record<string, Plan>): Promise<void> {
  await writeFile(getPlansPathIn(memoryDir), JSON.stringify(plans, null, 2), "utf-8");
}

async function savePlans(plans: Record<string, Plan>): Promise<void> {
  await savePlansIn(getMemoryDir(), plans);
}

export async function listPlansIn(memoryDir: string): Promise<Plan[]> {
  const plans = await loadPlansIn(memoryDir);
  return Object.values(plans).sort((a, b) => b.createdAt - a.createdAt);
}

async function listPlans(): Promise<Plan[]> {
  return listPlansIn(getMemoryDir());
}

export async function createPlanIn(memoryDir: string, title: string, description: string): Promise<Plan> {
  const id = generateId("plan");
  const plan: Plan = {
    id,
    title,
    description,
    conversationIds: [],
    createdAt: Date.now(),
  };
  const plans = await loadPlansIn(memoryDir);
  plans[id] = plan;
  await savePlansIn(memoryDir, plans);
  return plan;
}

async function createPlan(title: string, description: string): Promise<Plan> {
  return createPlanIn(getMemoryDir(), title, description);
}

export async function updatePlanIn(
  memoryDir: string,
  planId: string,
  updates: { title?: string; description?: string }
): Promise<Plan | null> {
  const plans = await loadPlansIn(memoryDir);
  const plan = plans[planId];
  if (!plan) return null;
  if (updates.title !== undefined) plan.title = updates.title;
  if (updates.description !== undefined) plan.description = updates.description;
  await savePlansIn(memoryDir, plans);
  return plan;
}

async function updatePlan(planId: string, updates: { title?: string; description?: string }): Promise<Plan | null> {
  return updatePlanIn(getMemoryDir(), planId, updates);
}

export async function deletePlanIn(memoryDir: string, planId: string): Promise<void> {
  const plans = await loadPlansIn(memoryDir);
  if (planId in plans) {
    delete plans[planId];
    await savePlansIn(memoryDir, plans);
  }
}

async function deletePlan(planId: string): Promise<void> {
  return deletePlanIn(getMemoryDir(), planId);
}

export async function addConversationToPlanIn(
  memoryDir: string,
  planId: string,
  conversationId: string
): Promise<Plan | null> {
  const plans = await loadPlansIn(memoryDir);
  const plan = plans[planId];
  if (!plan) return null;
  if (!plan.conversationIds.includes(conversationId)) {
    plan.conversationIds.push(conversationId);
    await savePlansIn(memoryDir, plans);
  }
  return plan;
}

async function addConversationToPlan(planId: string, conversationId: string): Promise<Plan | null> {
  return addConversationToPlanIn(getMemoryDir(), planId, conversationId);
}

export async function removeConversationFromPlanIn(
  memoryDir: string,
  planId: string,
  conversationId: string
): Promise<Plan | null> {
  const plans = await loadPlansIn(memoryDir);
  const plan = plans[planId];
  if (!plan) return null;
  plan.conversationIds = plan.conversationIds.filter((id) => id !== conversationId);
  await savePlansIn(memoryDir, plans);
  return plan;
}

async function removeConversationFromPlan(planId: string, conversationId: string): Promise<Plan | null> {
  return removeConversationFromPlanIn(getMemoryDir(), planId, conversationId);
}

export function registerPlansHandlers(): void {
  ipcMain.handle("plans:list", () => listPlans());
  ipcMain.handle("plans:create", (_e, title: string, description: string) => createPlan(title, description));
  ipcMain.handle("plans:update", (_e, planId: string, updates: { title?: string; description?: string }) =>
    updatePlan(planId, updates)
  );
  ipcMain.handle("plans:delete", (_e, planId: string) => deletePlan(planId));
  ipcMain.handle("plans:addConversation", (_e, planId: string, conversationId: string) =>
    addConversationToPlan(planId, conversationId)
  );
  ipcMain.handle("plans:removeConversation", (_e, planId: string, conversationId: string) =>
    removeConversationFromPlan(planId, conversationId)
  );
}

File: src-tauri/src/plans.rs
use std::collections::HashMap;
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};

use crate::env_util::generate_id;
use crate::memory::{AppState, PLANS_FILE};
use crate::paths::get_app_state_dir;
use crate::storage::{atomic_write_utf8, file_exists, read_json_object_file};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Plan {
    pub id: String,
    pub title: String,
    pub description: String,
    pub conversation_ids: Vec<String>,
    pub created_at: i64,
}

fn plans_path(memory_dir: &Path) -> PathBuf {
    memory_dir.join(PLANS_FILE)
}

pub async fn load_plans_in(
    _state: &AppState,
    memory_dir: &Path,
) -> Result<HashMap<String, Plan>, std::io::Error> {
    let path = plans_path(memory_dir);
    if !file_exists(&path).await {
        return Ok(HashMap::new());
    }
    let parsed = read_json_object_file(&path).await;
    let mut out = HashMap::new();
    if let Some(obj) = parsed.value.as_object() {
        for (id, value) in obj {
            if let Ok(plan) = serde_json::from_value::<Plan>(value.clone()) {
                out.insert(id.clone(), plan);
            }
        }
    }
    Ok(out)
}

async fn load_plans(state: &AppState) -> Result<HashMap<String, Plan>, std::io::Error> {
    load_plans_in(state, &get_app_state_dir()).await
}

pub async fn save_plans_in(
    state: &AppState,
    memory_dir: &Path,
    plans: &HashMap<String, Plan>,
) -> Result<(), std::io::Error> {
    let path = plans_path(memory_dir);
    let value = serde_json::to_value(plans).unwrap_or_else(|_| serde_json::json!({}));
    let pretty = serde_json::to_string_pretty(&value).unwrap_or_else(|_| "{}".into());
    atomic_write_utf8(&state.write_chains, &path, &pretty).await
}

pub async fn list_plans(state: &AppState) -> Result<Vec<Plan>, std::io::Error> {
    let plans = load_plans(state).await?;
    let mut rows: Vec<Plan> = plans.into_values().collect();
    rows.sort_by(|a, b| b.created_at.cmp(&a.created_at));
    Ok(rows)
}

pub async fn create_plan(
    state: &AppState,
    title: &str,
    description: &str,
) -> Result<Plan, std::io::Error> {
    let memory_dir = get_app_state_dir();
    let id = generate_id("plan");
    let plan = Plan {
        id: id.clone(),
        title: title.to_string(),
        description: description.to_string(),
        conversation_ids: vec![],
        created_at: chrono::Utc::now().timestamp_millis(),
    };
    let mut plans = load_plans_in(state, &memory_dir).await?;
    plans.insert(id, plan.clone());
    save_plans_in(state, &memory_dir, &plans).await?;
    Ok(plan)
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PlanUpdates {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
}

pub async fn update_plan(
    state: &AppState,
    plan_id: &str,
    updates: PlanUpdates,
) -> Result<Option<Plan>, std::io::Error> {
    let memory_dir = get_app_state_dir();
    let mut plans = load_plans_in(state, &memory_dir).await?;
    let Some(plan) = plans.get_mut(plan_id) else {
        return Ok(None);
    };
    if let Some(title) = updates.title {
        plan.title = title;
    }
    if let Some(description) = updates.description {
        plan.description = description;
    }
    let updated = plan.clone();
    save_plans_in(state, &memory_dir, &plans).await?;
    Ok(Some(updated))
}

pub async fn delete_plan(state: &AppState, plan_id: &str) -> Result<(), std::io::Error> {
    let memory_dir = get_app_state_dir();
    let mut plans = load_plans_in(state, &memory_dir).await?;
    if plans.remove(plan_id).is_some() {
        save_plans_in(state, &memory_dir, &plans).await?;
    }
    Ok(())
}

pub async fn add_conversation_to_plan(
    state: &AppState,
    plan_id: &str,
    conversation_id: &str,
) -> Result<Option<Plan>, std::io::Error> {
    let memory_dir = get_app_state_dir();
    let mut plans = load_plans_in(state, &memory_dir).await?;
    let Some(plan) = plans.get_mut(plan_id) else {
        return Ok(None);
    };
    if !plan.conversation_ids.contains(&conversation_id.to_string()) {
        plan.conversation_ids.push(conversation_id.to_string());
        let updated = plan.clone();
        save_plans_in(state, &memory_dir, &plans).await?;
        return Ok(Some(updated));
    }
    Ok(Some(plan.clone()))
}

pub async fn remove_conversation_from_plan(
    state: &AppState,
    plan_id: &str,
    conversation_id: &str,
) -> Result<Option<Plan>, std::io::Error> {
    let memory_dir = get_app_state_dir();
    let mut plans = load_plans_in(state, &memory_dir).await?;
    let Some(plan) = plans.get_mut(plan_id) else {
        return Ok(None);
    };
    plan.conversation_ids
        .retain(|id| id != conversation_id);
    let updated = plan.clone();
    save_plans_in(state, &memory_dir, &plans).await?;
    Ok(Some(updated))
}
